umbrella-feeds/src/lib.rs

```diff
@@ -0,0 +1,732 @@
+    // `signers`, but not the `message` here.
+    let param: UpdateParamsPartial = cursor.get()?;
+
+    // The input parameter is `UpdateParams` but we have only read the initial part
+    // of it with `UpdateParamsPartial` so far. We read in the `message` now.
+    // `(cursor.size() - cursor.cursor_position())` is the length of the message in
+    // bytes.
+    let mut message_bytes = vec![0; (cursor.size() - cursor.cursor_position()) as usize];
+
+    cursor.read_exact(&mut message_bytes)?;
+
+    // The message signed in the Concordium browser wallet is prepended with the
+    // `account` address and 8 zero bytes. Accounts in the Concordium browser wallet
+    // can either sign a regular transaction (in that case the prepend is
+    // `account` address and the nonce of the account which is by design >= 1)
+    // or sign a message (in that case the prepend is `account` address and 8 zero
+    // bytes). Hence, the 8 zero bytes ensure that the user does not accidentally
+    // sign a transaction. The account nonce is of type u64 (8 bytes).
+    let mut msg_prepend = vec![0u8; 32 + 8];
+
+    let vec_length = param.signers_and_signatures.len();
+
+    let mut message_hashes: Vec<[u8; 32]> = Vec::with_capacity(vec_length);
+
+    for i in 0..vec_length {
+        let signer = param.signers_and_signatures[i].0;
+
+        // Prepend the `account` address of the signer.
+        msg_prepend[0..32].copy_from_slice(signer.as_ref());
+        // Calculate the message hash.
+        message_hashes.push(
+            crypto_primitives
+                .hash_sha2_256(&[&msg_prepend[0..40], &message_bytes].concat())
+                .0,
+        );
+    }
+
+    Ok(message_hashes)
+}
+
+/// Helper function to verify the signature.
+/// This function throws if the signatures are not valid.
+#[receive(
+    contract = "umbrella_feeds",
+    name = "verifySignatures",
+    parameter = "UpdateParams",
+    crypto_primitives,
+    mutable
+)]
+fn verify_signatures<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    crypto_primitives: &impl HasCryptoPrimitives,
+) -> Result<(), CustomContractError> {
+    let param: UpdateParams = ctx.parameter_cursor().get()?;
+
+    ensure!(
+        param.signers_and_signatures.len() >= host.state().required_signatures as usize,
+        CustomContractError::NotEnoughSignatures
+    );
+
+    let mut prev_signer: Option<AccountAddress> = None;
+
+    let message_hashes = contract_view_message_hash(ctx, host, crypto_primitives)?;
+
+    let required_signatures = host.state().required_signatures;
+
+    let mut validators: Vec<AccountAddress> = Vec::with_capacity(required_signatures as usize);
+
+    // To save gas we check only required number of signatures.
+    // The case, where you can have part of signatures invalid but still enough valid in total is not supported.
+    for i in 0..required_signatures {
```

Comment:
This is not how we typically do this. All signatures must be valid and there has to be enough of them.

Is there a reason to deviate from that common logic?

Reply:
I agree with you that it is not efficient to provide additional signatures as input above the `required_signatures`.

@DZariusz Could jump in on this question.

Reply:
this is gas optimisation. If we require 6 signatures and we got 18, then after verifying 6 of them this is enough to consider data valid and save it. So we don't want to consume more gas if we already met the criteria.

Reply:
I added this comment:

```
We want to record all validators who submit (valid) signatures in a trustless/transparent way in the smart contract (to e.g. reward off-chain all validators for good behavior) that is why the smart contract allows submitting more signatures than the `required_signatures` here.
```